pkg/osbuild/librepo_source.go

```diff
@@ -0,0 +1,120 @@
+package osbuild
+
+import (
+	"fmt"
+
+	"github.com/osbuild/images/pkg/rpmmd"
+)
+
+// LibrepoSource wraps the org.osbuild.librepo osbuild source
+type LibrepoSource struct {
+	Items   map[string]*librepoSourceItem `json:"items"`
+	Options *librepoSourceOptions         `json:"options"`
+}
+
+func NewLibrepoSource() *LibrepoSource {
+	return &LibrepoSource{
+		Items: make(map[string]*librepoSourceItem),
+		Options: &librepoSourceOptions{
+			Mirrors: make(map[string]*librepoSourceMirror),
+		},
+	}
+}
+
+// AddPackage adds the given *depsolved* pkg to the downloading. It
+// needs the *depsovled* repoConfig so that the repoID of the two can
+// be matched up
+func (source *LibrepoSource) AddPackage(pkg rpmmd.PackageSpec, repos []rpmmd.RepoConfig) error {
+	pkgRepo := findRepoById(repos, pkg.RepoID)
+	if pkgRepo == nil {
+		return fmt.Errorf("cannot find repo-id %v for %v in %+v", pkg.RepoID, pkg.Name, repos)
+	}
+	if _, ok := source.Options.Mirrors[pkgRepo.Id]; !ok {
+		mirror, err := mirrorFromRepo(pkgRepo)
+		if err != nil {
+			return err
+		}
+		source.Options.Mirrors[pkgRepo.Id] = mirror
+	}
+	mirror := source.Options.Mirrors[pkgRepo.Id]
+	// XXX: should we error here if one package requests IgnoreSSL
+	// and one does not for the same mirror?
```

Comment:
This shouldn't be possible if the response from `osbuild-depsolve-dnf` hasn't been modified, right?  The ignore ssl option is per-repo so it shouldn't be inconsistent unless something was modified.  If that's the case, maybe an error is a good idea.

Reply:
Yeah, this should never happen in practise but it would be a potential security issue so I agree, lets check it.

Reply:
Done